### DISCUSSION TITLE
[Question] I have a question regarding the Authentication

### DISCUSSION THREAD

**Discussion by Rikthepixel** [SteamRE/SteamKit#925] [General]
Hello, my name is Rik and i have a few questions regarding the Authentication and all that jazz. im not a very experienced programmer so please keep that in mind moving forward.

I have had succes with login into a steam guard protected account with a bot made with SteamKit2. I have now implemented the stuff needed for the sentry file. The stupid thing is is that when i login once on my steambot (and create the sentryfile ) and then reopen it my program still asks for a 2FA code. 
It gives the EResult: AccountLoginDeniedNeedTwoFactor
and after inserting my VALID steam 2FA code it gives the following EResult:
Invalid

I dont know what "invalid" is supposed to mean. My bot is still able to connect to steam and do it's thing, but it still spits out "invalid".

So in summary I would like to request help with the following:
How do I make sure that when a valid authentication code is entered that steam does it's hash stuff and makes the bot be able to reconnect without me having to manually enter a steam SMA authcode again.

UserActivities.cs is where i handle my login related stuff. I would love if you guys were able to help me out and if you need more information please ask!

Thank you all a lot! And thanks to the Devs for making this awesome tool!

[UserActivities.zip](https://github.com/SteamRE/SteamKit/files/5361865/UserActivities.zip)

**Comment by azuisleet** [comment]
Since two-factor and steam guard are involved, every time you log in you are creating a new authentication session. 

Once you complete a challenge and authenticate, you will receive a SteamUser.LoginKeyCallback after signing in. You need to capture, acknowledge (SteamUser.AcceptNewLoginKey) and send that the next time you attempt to log in. If your authentication session is still valid, you won't have to complete a two-factor challenge.

**Comment by Rikthepixel** [comment]
thank you kindly, azuisleet for responding so quickly!

> Since two-factor and steam guard are involved, every time you log in you are creating a new authentication session.
> 
> Once you complete a challenge and authenticate, you will receive a SteamUser.LoginKeyCallback after signing in.  You need to capture, acknowledge (SteamUser.AcceptNewLoginKey) and send that the next time you attempt to log  in. If your authentication session is still valid, you won't have to complete a two-factor challenge.

Alright so i guess you mean kindof like this? As you can see im not very experienced :/
![afbeelding](https://user-images.githubusercontent.com/63408969/95690450-271ea400-0c18-11eb-8b49-b67c86ddea02.png)

Okey so once i have recieved the code and stored it. I know how to make a c# program read a file, but how do i send it to steam? Ive tried to look for what how i could send a key with that kind of format. Best bet i have is the SteamUser.MachineAuthDetails, but that seem to work. Honnestly you can call me clueless :/

it gives me a key in this kind of format:
(not the actual key) qCEUgydyuiAqLt0cbHU

btw do i still need to do something with that sentry file thingy? because it seems like that is not needed if login in with the steam guard code already works without it. also the sentry file doesnt really do that update thingy each time i login with the SMA

UPDATE: Never mind, i by accedent stumbled upon this gem: https://github.com/SteamRE/SteamKit/issues/387. The comment from  Netshroud commented on 27 Apr 2017 helped me out. It turns out that i can enter the Key into the SteamUser.LogOnDetails part.

UPDATE UPDATE: I was able to get the key, but sending it along with the other login details (and shouldRememberpassword set to true) doesnt seem to give me any positive result, it would seem that the EResult gives me a "InvalidPassword" or something in that nature.

**Comment by yaakov-h** [comment]
So all ok now then?

**Comment by Rikthepixel** [comment]
> 
> 
> So all ok now then?

Hey yaakoc-h, 

I got it kindof working, was able to get the key, but when I send the LoginKey (with the Rememberpassword stuff set to true) it give me an error that the password is invalid. yesterday i was digging for solutions to this problem and there was this one guy that got it working, but he didnt say what the problem was.

Anyways, Yes i am able to get the key. No, logging in without a steam Mobile Authenticator code was sadly not possible. So I stored the Key in a textfile and i am able to get the key in a second session.

maybe this is a better question: "What does the LoginKey do?"

again thanks for you guy's help!
